plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Currency.hs

```diff
@@ -0,0 +1,171 @@
+                    --   the currency is forged.
+                    , curRefAmount            :: Int
+                    -- ^ How much of the currency can be forged
+                    , curRefTokenName         :: TokenName
+                    -- ^ Token name
+                    }
+
+P.makeLift ''Currency
+
+mkCurrency :: TxOutRef -> Int -> String -> Currency
+mkCurrency (TxOutRefOf h i) amt n = 
+    Currency
+        { curRefTransactionOutput = (V.plcTxHash h, i)
+        , curRefAmount            = amt
+        , curRefTokenName         = mkTokenName n
+        }
+
+curValidator :: Currency -> ValidatorScript
+curValidator cur =
+    ValidatorScript (Ledger.applyScript mkValidator (Ledger.lifted cur)) where
+        mkValidator = Ledger.fromCompiledCode ($$(P.compile [||
+            let validate :: Currency -> () -> () -> V.PendingTx -> ()
+                validate (Currency (refHash, refIdx) amt token) () () p = 
+                    let 
+                        ownSymbol = $$(V.ownCurrencySymbol) p
+                        forged = $$(V.valueForged) p
+                        expected = $$(Value.singleton) ownSymbol token amt 
+
+                        -- True if the pending transaction forges the amount of
+                        -- currency that we expect
+                        forgeOK =
+                            let v = $$(Value.eq) expected forged
+                            in $$(P.traceIfFalseH) "Value forged different from expected" v
+
+                        -- True if the pending transaction spends the output
+                        -- identified by @(refHash, refIdx)@
+                        txOutputSpent = 
+                            let v = $$(V.spendsOutput) p refHash refIdx
+                            in  $$(P.traceIfFalseH) "Pending transaction does not spend the designated transaction output" v
+
+                    in
+                        if $$(P.and) forgeOK txOutputSpent
+                        then ()
+                        else $$(P.error) ($$(P.traceH) "Invalid forge" ())
+            in
+                validate
+            ||]))
+
+-- | Currency data script (unit value).
+curDataScript :: DataScript
+curDataScript = DataScript $ Ledger.lifted ()
+
+-- | Currency redeemer (unit value).
+curRedeemer :: RedeemerScript
+curRedeemer = RedeemerScript $ Ledger.lifted ()
+
+-- | The address of a 'Currency' contract.
+curAddress :: Currency -> Address
+curAddress = Ledger.scriptAddress . curValidator
+
+-- | The 'Value' forged by the 'curValidator' contract
+forgedValue :: Currency -> Value
+forgedValue cur = 
+    let a = plcCurrencySymbol (curAddress cur)
+        t = curRefTokenName cur
+        i = curRefAmount cur
+    in
+        $$(Value.singleton) a t i
+
+-- | @forge c n@ forges @n@ units of a currency called @c@ and
+--   pays them to a public key address owned by the wallet.
+forge :: (WalletAPI m, WalletDiagnostics m) => String -> Int -> m Currency
+forge nm amount = do

```

Comment:
This feels rather like hard work but I don't have any particular suggestions for how to make it easier atm.

Reply:
It seems like we probably want to provide a more abstract forging thing like the `payToPubKey` function, but it should allow a more flexible specification including the MPS, fungibility etc.